phac-nml/irida-next
# Workflow DB Representation

## What does this PR do and why?
_Describe in detail what your merge request does and why._
Fixes #245

## Screenshots or screen recordings
_Screenshots are required for UI changes, and strongly recommended for all other pull requests._
N/A

## How to set up and validate locally
_Numbered steps to set up and validate the change are strongly suggested._
Clean and migrate db
```bash
./scripts/clean_dev.sh
```
Check that tables are generated and can be interacted with
```bash
rails c
```
```ruby
# *****************************
# Test tables
# *****************************
# workflow_executions and samples_workflow_executions should exist
ActiveRecord::Base.connection.tables

# *****************************
# Test workflow_executions
# *****************************
# create a new workflow execution, without the metadata object
we = WorkflowExecution.create submitter: User.find(1), run_id: "my run ID"
we.valid? # should return true

# See our Workflow Execution in the db
WorkflowExecution.all

# *****************************
# Test samples_workflow_executions
# *****************************
swe = SamplesWorkflowExecution.create(sample: Sample.find(1), workflow_execution: WorkflowExecution.find(1), samplesheet_params: { my_key: 'myvalue' })
# Verify that the above command inserts the `samplessheet_params` hash as jsonb. This can be seen in the TRANSACTION event.
swe.valid? # check validity
SampleWorkflowExecution.all # view it via db

# *****************************
# Test join table linking
 *******************************

# grab out sample and workflow execution
s1 = Sample.find(1)
we1 = WorkflowExecution.find(1)

# get SamplesWorkflowExecutions via Sample
s1.samples_workflow_executions

# get WorkflowExecutions via Sample
s1.workflow_executions

# get SamplesWorkflowExecutions via WorkflowExecution
we1.samples_workflow_executions

# get Samples via WorkflowExecution
we1.samples

```
## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.
